### PR TITLE
Adds TLS Termination Support

### LINKED ISSUE #471: Support simple TLS termination
*Description*:
Add support for HTTPS/TLS termination
Examples shown here https://gateway-api.sigs.k8s.io/v1alpha2/guides/tls/#examples

Looks like most of the work is already done in gateway-api translator https://github.com/envoyproxy/gateway/blob/8e1e9815a880236d49df1d21735f1a41fbb9b3d6/internal/gatewayapi/translator.go#L52 and the provider simply needs to reconcile secrets and pass it over.

### PR DESCRIPTION
**Pull request by danehans**
Adds support for TLS termination.

Fixes #471

### CODE REVIEW

File: internal/gatewayapi/translator.go [hunk 1]
```diff
@@ -486,7 +486,7 @@ func (t *Translator) ProcessListeners(gateways []*GatewayContext, xdsIR XdsIRMap
 					proto = ir.HTTPSProtocolType
 				}
 				infraPort := ir.ListenerPort{
-					Name:          irInfraPortName(listener),
+					Name:          string(listener.Name),
```
Comment (arkodg) [inline internal/gatewayapi/translator.go:489]: looks like you need to rebase

**Comment by arkodg** [comment]
https://github.com/envoyproxy/gateway/blob/b8f234b463833b56f554e993faac19018464c1e2/internal/gatewayapi/runner/runner.go#L82 will need to be enhanced with a function that scrubs outs the key material, deleting it in this PR and raising a issue to track it should unblock this PR.

File: internal/gatewayapi/translator.go (hunk 1 above)
Comment (danehans) [reply internal/gatewayapi/translator.go]: ack

**Comment by danehans** [comment]
@arkodg regagarding https://github.com/envoyproxy/gateway/pull/519#issuecomment-1272178552, removed.

File: internal/provider/kubernetes/gateway.go [hunk 2]
```diff
@@ -280,6 +413,117 @@ func (r *gatewayReconciler) envoyServiceForGateway(ctx context.Context, gateway
 	return svc, nil
 }
 
+// otherGatewaysRefSecret returns an error if a managed Gateway other than the provided
+// gateway references the provided secret, or if an error is encountered while checking.
+func (r *gatewayReconciler) otherGatewaysRefSecret(ctx context.Context, gateway *gwapiv1b1.Gateway, secret *corev1.Secret) error {
```
Comment (arkodg) [inline internal/provider/kubernetes/gateway.go:418]: nit returning (bool, err) would be clearer from a client perspective

File: internal/provider/kubernetes/gateway.go [hunk 3]
```diff
@@ -179,6 +286,8 @@ func (r *gatewayReconciler) Reconcile(ctx context.Context, request reconcile.Req
 	}
 
 	found := false
+	var secrets []corev1.Secret
+	var requestedGateway gwapiv1b1.Gateway
```
Comment (arkodg) [inline internal/provider/kubernetes/gateway.go:290]: I dont quite understand how `requestedGateway` is being used below in the `found` & `!found` logic

File: internal/provider/kubernetes/gateway.go [hunk 4]
```diff
@@ -280,6 +413,117 @@ func (r *gatewayReconciler) envoyServiceForGateway(ctx context.Context, gateway
...
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// secretsForGateway returns the Secrets referenced by the provided gateway listeners.
+func (r *gatewayReconciler) secretsForGateway(ctx context.Context, gateway *gwapiv1b1.Gateway) ([]corev1.Secret, error) {
```
Comment (arkodg) [inline internal/provider/kubernetes/gateway.go:445]: would help if there was a test for `secretsForGateway` added in this or follow up PRs

File: internal/envoygateway/scheme.go [hunk 5]
```diff
@@ -3,6 +3,7 @@ package envoygateway
 import (
 	"k8s.io/apimachinery/pkg/runtime"
 	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	gwapiv1a2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
```
Comment (zirain) [inline internal/envoygateway/scheme.go:6]: curious why need this?

File: internal/provider/kubernetes/gateway.go [hunk 6]
```diff
@@ -280,6 +437,124 @@ func (r *gatewayReconciler) envoyServiceForGateway(ctx context.Context, gateway
...
+							opts := client.ListOptions{Namespace: string(*ref.Namespace)}
+							if err := r.client.List(ctx, refGrants, &opts); err != nil {
+								return nil, nil, fmt.Errorf("error listing referencegrants")
+							}
+							var gwRefd, secretRefd bool
+							for _, rg := range refGrants.Items {
+								for _, from := range rg.Spec.From {
+									if from.Group == gwapiv1a2.GroupName &&
+										from.Kind == gatewayapi.KindGateway {
+										gwRefd = true
```
Comment (arkodg) [inline internal/provider/kubernetes/gateway.go:493]: break ?

File: internal/provider/kubernetes/gateway.go [hunk 7]
```diff
@@ -280,6 +437,124 @@ func (r *gatewayReconciler) envoyServiceForGateway(ctx context.Context, gateway
...
+									}
+								}
+								for _, to := range rg.Spec.To {
+									if to.Group == corev1.GroupName &&
+										to.Kind == gatewayapi.KindSecret {
+										if to.Name == nil || *to.Name == gwapiv1a2.ObjectName(ref.Name) {
+											secretRefd = true
```
Comment (arkodg) [inline internal/provider/kubernetes/gateway.go:500]: break ?

File: internal/provider/kubernetes/gateway.go [hunk 8]
```diff
@@ -280,6 +437,124 @@ func (r *gatewayReconciler) envoyServiceForGateway(ctx context.Context, gateway
...
+// secretsAndRefGrantsForGateway returns the Secrets referenced by the provided gateway listeners.
+// If the provided Gateway references a Secret in a different namespace, a list of
+// ReferenceGrants is returned that permit the cross namespace Secret reference.
+func (r *gatewayReconciler) secretsAndRefGrantsForGateway(ctx context.Context, gateway *gwapiv1b1.Gateway) ([]corev1.Secret, []gwapiv1a2.ReferenceGrant, error) {
+	var secrets []corev1.Secret
+	var returnedGrants []gwapiv1a2.ReferenceGrant
+	if terminatesTLS(gateway) {
+		for i := range gateway.Spec.Listeners {
+			listener := gateway.Spec.Listeners[i]
+			if listener.TLS != nil {
```
Comment (arkodg) [inline internal/provider/kubernetes/gateway.go:477]: `terminatesTLS` seems to loop across all listeners, and so does this line, do you think `terminatesTLS(gateway)` can be refined to `terminatesTLS(listener)` here ?

**Review by arkodg — DISMISSED** [review]

**Review by arkodg — APPROVED** [review]

**Pull request merged**